anitya/lib/backends/crates.py

```diff
@@ -0,0 +1,216 @@
+        "required": [
+          "crate",
+          "created_at",
+          "dl_path",
+          "downloads",
+          "features",
+          "id",
+          "links",
+          "num",
+          "updated_at",
+          "yanked"
+        ]
+      }
+    },
+    "items": {
+      "anyOf": [
+        { "$ref": "#/definitions/version" }
+      ]
+    }
+  }
+
+For example, ``https://crates.io/api/v1/crates/itoa/versions`` results in::
+
+    {
+        "versions": [
+            {
+                "crate": "itoa",
+                "created_at": "2017-01-26T05:02:29Z",
+                "dl_path": "/api/v1/crates/itoa/0.2.1/download",
+                "downloads": 150,
+                "features": {},
+                "id": 43354,
+                "links": {
+                    "authors": "/api/v1/crates/itoa/0.2.1/authors",
+                    "dependencies": "/api/v1/crates/itoa/0.2.1/dependencies",
+                    "version_downloads": "/api/v1/crates/itoa/0.2.1/downloads"
+                },
+                "num": "0.2.1",
+                "updated_at": "2017-01-26T05:02:29Z",
+                "yanked": false
+            }
+    }
+"""
+import requests
+
+from anitya.lib.backends import BaseBackend
+from anitya.lib.exceptions import AnityaPluginException
+
+
+class CratesBackend(BaseBackend):
+    """The crates class for projects hosted on crates.io."""
+
+    name = 'crates.io'
+    examples = [
+        'https://crates.io/crates/clap',
+        'https://crates.io/crates/serde',
+    ]
+
+    @classmethod
+    def _get_versions(cls, project):
+        """
+        Make a request for all versions of the project provided.
+
+        Args:
+            project (anitya.lib.model.Project): The Rust project to retrieve
+                versions for.
+
+        Returns:
+            list: A list of version dictionaries. Each dictionary is in the format::
+                {
+                    "crate": "itoa",
+                    "created_at": "2017-01-26T05:02:29Z",
+                    "dl_path": "/api/v1/crates/itoa/0.2.1/download",
+                    "downloads": 150,
+                    "features": {},
+                    "id": 43354,
+                    "links": {
+                        "authors": "/api/v1/crates/itoa/0.2.1/authors",
+                        "dependencies": "/api/v1/crates/itoa/0.2.1/dependencies",
+                        "version_downloads": "/api/v1/crates/itoa/0.2.1/downloads"
+                    },
+                    "num": "0.2.1",
+                    "updated_at": "2017-01-26T05:02:29Z",
+                    "yanked": false
+                }
+
+        Raises:
+            AnityaPluginException: If the URL was unreachable or the response
+                was in an unexpected format.
+        """
+        url = 'https://crates.io/api/v1/crates/{}/versions'.format(project.name)
+        try:
+            req = cls.call_url(url)
+            req.raise_for_status()
+            data = req.json()
```

Comment:
requests hands the data to the ``json`` module and it raises the exception. In Python 2 this will result in a ValueError (handled below) and in Python 3 it'll be a [JSONDecodeError](https://docs.python.org/3.6/library/json.html#json.JSONDecodeError) which is a subclass of ValueError so it'll also get handled below.